pkg/verification/service/verification_service.go

```diff
@@ -0,0 +1,274 @@
+package service
+
+import (
+	"crypto/md5"
+	"crypto/rand"
+	"encoding/hex"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/kevinburke/twilio-go"
+
+	"github.com/codeready-toolchain/registration-service/pkg/application/service"
+	"github.com/codeready-toolchain/registration-service/pkg/application/service/base"
+	servicecontext "github.com/codeready-toolchain/registration-service/pkg/application/service/context"
+
+	"github.com/codeready-toolchain/registration-service/pkg/errors"
+
+	"github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"
+	"github.com/codeready-toolchain/registration-service/pkg/log"
+	"github.com/gin-gonic/gin"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+)
+
+const (
+	codeCharset = "0123456789"
+	codeLength  = 6
+
+	TimestampLayout = "2006-01-02T15:04:05.000Z07:00"
+)
+
+// ServiceConfiguration represents the config used for the verification service.
+type ServiceConfiguration interface {
+	GetTwilioAccountSID() string
+	GetTwilioAuthToken() string
+	GetTwilioFromNumber() string
+	GetVerificationMessageTemplate() string
+	GetVerificationAttemptsAllowed() int
+	GetVerificationDailyLimit() int
+	GetVerificationCodeExpiresInMin() int
+}
+
+// ServiceImpl represents the implementation of the verification service.
+type ServiceImpl struct {
+	base.BaseService
+	config     ServiceConfiguration
+	HttpClient *http.Client
+}
+
+type VerificationServiceOption func(svc *ServiceImpl)
+
+// NewVerificationService creates a service object for performing user verification
+func NewVerificationService(context servicecontext.ServiceContext, cfg ServiceConfiguration, opts ...VerificationServiceOption) service.VerificationService {
+	s := &ServiceImpl{
+		BaseService: base.NewBaseService(context),
+		config:      cfg,
+	}
+
+	for _, opt := range opts {
+		opt(s)
+	}
+	return s
+}
+
+// InitVerification sends a verification message to the specified user, using the Twilio service.  If successful,
+// the user will receive a verification SMS.  The UserSignup resource is updated with a number of annotations in order
+// to manage the phone verification process and protect against system abuse.
+func (s *ServiceImpl) InitVerification(ctx *gin.Context, userID string, e164PhoneNumber string) error {
+	var initError error
```

Comment:
i'm in favour or just returning the error rather than re-assigning it and waiting until the end to return it. but in any case this is only used on line 144 the first time. maybe we could define it there instead?

Reply:
No problem, moved it down further in the code.